File: Data Structure/java/ThreeSum.java
import java.util.*;

/**
 * Given an array nums of n integers, are there elements a, b, c in nums such
 * that a + b + c = 0? Find all unique triplets in the array which gives the sum
 * of zero.
 * 
 * Idea of solution: we essentially need to find three numbers x, y, and z such
 * that they add up to the given value. If we fix one of the numbers say x, we
 * are left with the two-sum problem at hand!
 * 
 * Time Complexity: O(n^2), Space Complextiy: O(1)
 */

public class ThreeSum {
    /**
     * @param nums: an interger array with negative and duplicates.
     * @return a list of all unique triplets in the array with sum = 0
     */
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            // Two pointers to find solution
            int left = i + 1, right = nums.length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    res.add(new ArrayList<Integer>(Arrays.asList(nums[i], nums[left], nums[right])));
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1])
                        left++;
                    while (left < right && nums[right] == nums[right + 1])
                        right--;
                }
                if (sum > 0)
                    right--;
                if (sum < 0)
                    left++;
            }
        }
        return res;
    }
}

File: Data Structure/python/three_sum.py
from typing import List


class Solution:
    """
    Args:
        nums: Given an array of integers with negative value and duplicates.

    Returns:
        return all unique triplets in the array which gives the sum of zero.

    Idea of Solution:
       The idea is to sort an input array and then run through all indices of a possible first element of a triplet.
       For each possible first element we make a standard bi-directional 2Sum sweep of the remaining part of the array.
       Also we want to skip equal elements to avoid duplicates in the answer without making a set or smth like that.

    Time Complexity: O(n^2)
    Space Complexity: O(1)
    """

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # Sort first
        nums.sort()
        res = []
        for i in range(len(nums) - 2):
            # Break the loop if there is no possible solution. i.e. the min in array is greater than 0.
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i-1]:
                continue
            # Two pointer
            left, right = i + 1, len(nums) - 1
            while left < right:
                target = nums[i] + nums[left] + nums[right]
                if target == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    # Remove duplicates
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                if target > 0:
                    right -= 1
                if target < 0:
                    left += 1
        return res
